### PR TITLE
mysql: add MyDecimal data type.

### PR DESCRIPTION
**Pull request by coocood**
This PR only includes necessary method to make `FromString` method work. Which is used to write unit test.
More methods and test cases will be added in the following PR.

ported from 
https://github.com/mysql/mysql-server/blob/5.7/strings/decimal.c

### CODE REVIEW

**Comment by coocood** [comment]
@shenli @BusyJay @zimulala

File: mysql/mydecimal.go [hunk 1]
```diff
@@ -0,0 +1,776 @@
...
+	ten2 = 100
+	ten3 = 1000
+	ten4 = 10000
+	ten5 = 100000
+	ten6 = 1000000
+	ten7 = 10000000
+	ten8 = 100000000
+	ten9 = 1000000000
+
+	maxWordBufLen = 9 // a MyDecimal holds 9 words.
```
Comment (shenli) [inline mysql/mydecimal.go]: a -> A

File: mysql/mydecimal.go [hunk 2]
```diff
@@ -0,0 +1,776 @@
...
+)
+
+var (
+	wordBufLen    = 9
+	powers10      = [10]int32{ten0, ten1, ten2, ten3, ten4, ten5, ten6, ten7, ten8, ten9}
+	dig2bytes     = [10]int{0, 1, 1, 2, 2, 3, 3, 4, 4, 4}
+	zeroMyDecimal = MyDecimal{digitsInt: 1}
+)
+
+// add adds a and b and carry, returns new carry.
```
Comment (shenli) [inline mysql/mydecimal.go]: returns the sum and new carray

**Comment by shenli** [comment]
@tiancaiamao PTAL

File: mysql/mydecimal.go [hunk 3]
```diff
@@ -0,0 +1,776 @@
...
+			digitsInt = wordsInt * digitsPerWord
+		}
+	}
+	d.digitsInt = digitsInt
+	d.digitsFrac = digitsFrac
+	wordIdx := wordsInt
+	strIdxTmp := strIdx
+	var word int32
+	var innerIdx int
+	for digitsInt > 0 {
```
Comment (tiancaiamao) [inline mysql/mydecimal.go]: this code patten is redundancy with line 214~228, extract to a function maybe better?

File: mysql/mydecimal.go [hunk 4]
```diff
@@ -0,0 +1,776 @@
...
+  SYNOPSIS
+    digits_bounds()
+      from         - decimal number for processing
+      start - index (from 0 ) of first decimal digits.
+      end   - index of position just after last decimal digit.
+*/
+func (d *MyDecimal) digitBounds() (start, end int) {
+	var (
+		i int
+	)
```
Comment (tiancaiamao) [inline mysql/mydecimal.go]: var i int

File: mysql/util.go [hunk 5]
```diff
@@ -55,3 +60,54 @@ func GetDefaultDecimal(tp byte) int {
...
+func myMin(a, b int) int {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+// strToInt converts a string to an integer in best effort.
+// TODO: handle overflow and add unittest.
+func strToInt(str string) (int64, error) {
```
Comment (tiancaiamao) [inline mysql/util.go]: what's the difference with `strconv.ParseInt`

File: mysql/mydecimal_test.go [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+		input   string
+		output  uint64
+		errcode int
+	}{
+		{"12345", 12345, 0},
+		{"0", 0, 0},
+		/* ULLONG_MAX = 18446744073709551615ULL */
+		{"18446744073709551615", 18446744073709551615, 0},
+		{"18446744073709551616", 18446744073709551615, 2},
+		{"-1", 0, 2},
```
Comment (tiancaiamao) [inline mysql/mydecimal_test.go]: add `math.MinInt64` and `math. MinInt64-1`, in case that overflow not processed properly by the code

Comment (tiancaiamao) [inline mysql/mydecimal_test.go]: "-9223372036854775808" "-9223372036854775809"

File: mysql/mydecimal.go [hunk 7]
```diff
@@ -0,0 +1,776 @@
...
+	// A word is an int32 value can hold 9 digits.
+	// 0 <= word < wordBase
+	wordBuf [maxWordBufLen]int32
+}
+
+func (d *MyDecimal) String() string {
+	return fmt.Sprintf("di:%d, df:%d, neg:%v, buf:%v", d.digitsInt, d.digitsFrac, d.negative, d.wordBuf)
+}
+
+// FromString parse decimal from string.
```
Comment (shenli) [inline mysql/mydecimal.go]: parses

File: mysql/mydecimal.go [hunk 8]
```diff
@@ -0,0 +1,776 @@
...
+	for word%powers10[i] == 0 {
+		i++
+		trailing++
+	}
+	return trailing
+}
+
+// MyDecimal represents a decimal value.
+type MyDecimal struct {
+	// the number of *decimal* digits before the point.
```
Comment (shenli) [inline mysql/mydecimal.go]: the -> The

File: mysql/mydecimal.go [hunk 9]
```diff
@@ -0,0 +1,776 @@
...
+	digitsInt int
+
+	// the number of decimal digits after the point
```
Comment (shenli) [inline mysql/mydecimal.go]: the -> The
Add stop mark at the end.

File: mysql/mydecimal.go [hunk 10]
```diff
@@ -0,0 +1,776 @@
...
+		}
+		if ec != eDecOverflow {
+			ec = d.Shift(int(exponent))
+		}
+	}
+	return ec
+}
+
+// Shift shifts decimal digits in given number (with rounding if it need), shift > 0 means shift to left shift,
+// shift < 0 meand right shift. In fact it is multipling on 10^shift.
```
Comment (shenli) [inline mysql/mydecimal.go]: meand -> means

File: mysql/mydecimal.go [hunk 11]
```diff
@@ -0,0 +1,776 @@
...
+		/* index of position after last decimal digit */
+		digitEnd int
+		/* index of digit position just after point */
+		point = (d.digitsInt + digitsPerWord - 1) / digitsPerWord * digitsPerWord
+		/* new point position */
+		newPoint = point + shift
+		/* number of digits in result */
+		digitsInt, digitsFrac int
+		/* return code */
+		err      = eDecOK
```
Comment (shenli) [inline mysql/mydecimal.go]: err -> errCode

File: mysql/mydecimal.go [hunk 12]
```diff
@@ -0,0 +1,776 @@
...
+		newFront int
+	)
+	if shift == 0 {
```
Comment (shenli) [inline mysql/mydecimal.go]: Can we move this to the beginning?

File: mysql/mydecimal.go [hunk 13]
```diff
@@ -0,0 +1,776 @@
...
+	} else {
+		end = (bufEnd + 1) * digitsPerWord
+		i = 1
+	}
+	end -= countTrailingZeroes(i, d.wordBuf[bufEnd])
+	return start, end
+}
+
+/*
+  doMiniLeftShift do left shift for alignment of data in buffer
```
Comment (shenli) [inline mysql/mydecimal.go]: do -> does

File: mysql/mydecimal.go [hunk 14]
```diff
@@ -0,0 +1,776 @@
...
+//  eDecOK/eDecTruncated
+func (d *MyDecimal) Round(to *MyDecimal, frac int, mode roundMode) (errcode int) {
+	// wordsFracTo is the number of fraction words in buffer.
+	wordsFracTo := (frac + 1) / digitsPerWord
+	if frac > 0 {
+		wordsFracTo = (frac + digitsPerWord - 1) / digitsPerWord
+	}
+	wordsFrac := (d.digitsFrac + digitsPerWord - 1) / digitsPerWord
+	wordsInt := (d.digitsInt + digitsPerWord - 1) / digitsPerWord
+	//var x, y int32
```
Comment (shenli) [inline mysql/mydecimal.go]: useless code

File: mysql/mydecimal_test.go [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+
+package mysql
+
+import (
+	. "github.com/pingcap/check"
+)
+
+var _ = Suite(&testMyDecimalSuite{})
+
+type testMyDecimalSuite struct {
```
Comment (shenli) [inline mysql/mydecimal_test.go]: Any testcase for parse decimal with frac?

File: mysql/util.go (hunk 5 above)
Comment (coocood) [inline mysql/util.go]: The returned result is different when error is returned.

File: mysql/mydecimal.go (hunk 3 above)
Comment (coocood) [inline mysql/mydecimal.go]: Those to blocks of code has a few differences.
I think extracting those code to a function will make it harder to read.

**Comment by tiancaiamao** [comment]
LGTM
we should add more tests to clarify the behavior of  those functions later.

File: mysql/mydecimal.go [hunk 16]
```diff
@@ -0,0 +1,776 @@
...
+	if sum >= wordBase {
+		carry = 1
+		sum -= wordBase
+	} else {
+		carry = 0
+	}
+	return sum, carry
+}
+
+// fixWordCntError limits word count in wordBufLen, and returns overflow or truncate error
```
Comment (zimulala) [inline mysql/mydecimal.go]: Add . at the end of a line

File: mysql/mydecimal.go [hunk 17]
```diff
@@ -0,0 +1,776 @@
...
+type MyDecimal struct {
+	// the number of *decimal* digits before the point.
+	digitsInt int
+
+	// the number of decimal digits after the point
+	digitsFrac int
+
+	negative bool
+
+	// an array of int32 words.
```
Comment (zimulala) [inline mysql/mydecimal.go]: an -> An

File: mysql/mydecimal_test.go (hunk 15 above)
Comment (coocood) [inline mysql/mydecimal_test.go]: Those test case will be added in Next PR.

File: mysql/mydecimal.go [hunk 18]
```diff
@@ -0,0 +1,776 @@
...
+		}
+		digitShift := wordShift * digitsPerWord
+		digitBegin += digitShift
+		digitEnd += digitShift
+		newPoint += digitShift
+	}
+	/*
+	   If there are gaps then fill ren with 0.
+
+	   Only one of following 'for' loops will work becouse wordIdxBegin <= wordIdxEnd
```
Comment (zimulala) [inline mysql/mydecimal.go]: Add . at the end of a line, the same as line 397, line 417

File: mysql/mydecimal.go [hunk 19]
```diff
@@ -0,0 +1,776 @@
...
+	d.digitsFrac = digitsFrac
+	return err
+}
+
+/*
+  digitBounds returns bounds of decimal digits in the number
+
+  SYNOPSIS
+    digits_bounds()
+      from         - decimal number for processing
```
Comment (zimulala) [inline mysql/mydecimal.go]: It needs to align.

**Comment by coocood** [comment]
@zimulala PTAL

**Comment by shenli** [comment]
LGTM

**Pull request merged**